### PR TITLE
refactor!: typescript rewrite

### PR DESCRIPTION
**Pull request by kibibites**

### CODE REVIEW

**Comment by kibibites** [comment]
ill start rewriting tomorrow

**Comment by kibibites** [comment]
21 commits already

**Comment by kibibites** [comment]
1 more file :weary:

**Comment by kibibites** [comment]
```
i-spin ~ git version 2.34.0
---------------------------
Project: lambda-bot-js (2 branches)
HEAD: 9b375f8 (rewrite/typescript, origin/rewrite/typescript)
Created: 4 days ago
Language: TypeScript (100.0 %)
Dependencies: 15 (npm)
Authors: 81% i-spin 58
         17% sourtaste000 12
         1% i-spin 1
Last change: a minute ago
Contributors: 4
Repo: https://github.com/i-spin/lambda-bot-js.git
Commits: 72
Lines of code: 539
Size: 121.06 KiB (30 files)
License: Unlicense
```

**Comment by i-spin** [comment]
TODOs before merging:
- [ ] fix all eslint issues
- [ ] fix all typescript issues
- [ ] make sure every command work

**Comment by kibibites** [comment]
why doesnt gitkraken sign my commits

**Review by emyfops — COMMENTED** [review]
Everything looks good 👍

File: src/commands/capes.ts [hunk 1]
```diff
@@ -65,31 +64,28 @@ export default {
     interaction.options.data.forEach(async (index) => {
       switch (index.name) {
         case 'pull': {
-          const forced = interaction.options.getBoolean('force') ?? false;
+          const forced: boolean = interaction.options.getBoolean('force') ?? false;
+          if(forced === undefined) return await interaction.reply({ embeds: [embedUtils.error("error")]});
```
Comment (kibibites) [inline src/commands/capes.ts]: check is unneccasary

File: src/commands/capes.ts [hunk 2]
```diff
@@ -65,31 +64,28 @@ export default {
...
-          const user = interaction.options.getString('user_id')!.split("'")[0];
-          const minecraftUUID = uuidUtils.usernameToUUID(minecraftUsername);
+          const user = await discordUtils.checkuser(interaction.options.getString('user_id')!);
+          if(user === undefined) return interaction.reply({ embeds: [embedUtils.error('Please provide a valid user ID')]});
+          const minecraftUUID = await uuidUtils.usernameToUUID(minecraftUsername);
           if (!minecraftUUID) return interaction.reply({ embeds: [embedUtils.error('Invalid username or nonexistent player')] });
-          capeUtils.add(parseInt(user, 10), minecraftUUID);
-          return interaction.reply({ embeds: [embedUtils.success(`Added <@${user}>, Info: \`\`\`Username: ${minecraftUsername}\nUUID: ${minecraftUUID}\`\`\``)] });
+          await capeUtils.add(user.id, minecraftUUID);
+          return interaction.reply({ embeds: [embedUtils.success(`Added <@${user.id}>, Info: \`\`\`Username: ${minecraftUsername}\nUUID: ${minecraftUUID}\`\`\``)], allowedMentions: { "users": [] } });
```
Comment (kibibites) [inline src/commands/capes.ts]: I don't think mentions in embeds will ping someone

File: src/utils/capes.ts [hunk 3]
```diff
@@ -1,13 +1,13 @@
...
-import fs from 'graceful-fs';
+import * as fs from 'graceful-fs';
 import { capeRepo } from '../../config';
 
 const octokit = new Octokit({
   auth: process.env.GITHUB_TOKEN,
 });
 
-function pull(options?: { forced: boolean }) {
+async function pull(forced?: boolean) {
```
Comment (kibibites) [inline src/utils/capes.ts:10]: why did you make this async?

File: src/utils/capes.ts [hunk 4]
```diff
@@ -16,16 +16,18 @@ function pull(options?: { forced: boolean }) {
   })
     .then((res) => {
       const data = res.data as any;
-      if (fs.existsSync('../capes.json') && Buffer.from((res.data as any).content, 'base64') !== fs.readFileSync('../capes.json')) {
-        if (options?.forced) return;
-        throw Error('Conflict detected in capes.json');
+      if (fs.existsSync('../capes.json')) {
+        if (String(Buffer.from(data.content, "base64")) == String(fs.readFileSync('../capes.json'))) {
```
Comment (kibibites) [inline src/utils/capes.ts]: use singlequotes

File: src/utils/capes.ts [hunk 5]
```diff
@@ -16,16 +16,18 @@ function pull(options?: { forced: boolean }) {
...
+        fs.writeFileSync('../capes.json', Buffer.from(data.content, 'base64'));
+        fs.writeFileSync('../capes.json.shasum', data.sha);
       }
-      fs.writeFileSync('../capes.json', Buffer.from(data.content, 'base64'));
-      fs.writeFileSync('../capes.json.shasum', data.sha);
     });
 }
 
-function push() {
+async function push() {
```
Comment (kibibites) [inline src/utils/capes.ts:30]: same above, why make this async?

File: src/utils/capes.ts [hunk 6]
```diff
@@ -47,10 +49,10 @@ function push() {
   }).then((res) => res.data).catch((e) => e);
 }
 
-function add(discordId: number, uuid: string) {
+async function add(discordId: string, uuid: string) {
   const capes = JSON.parse(fs.readFileSync('../capes.json', 'utf8'));
   const template = {
-    id: discordId,
+    id: Number(discordId),
```
Comment (kibibites) [inline src/utils/capes.ts:55]: or use `parseInt()`, or just take a number type in the argument

File: src/utils/minecraftProfile.ts [hunk 7]
```diff
@@ -1,30 +1,22 @@
...
+import Names from '../types/names';
+import Textures, { TextureURL } from '../types/textures';
 
-function profile(name: string): Profile | undefined {
-  fetch(`https://api.mojang.com/users/profiles/minecraft/${name}`)
-    .then((res) => res.json())
-    .then((data) => (data as Profile))
-    .catch((_err) => false);
-  return undefined;
+async function profile(name: string): Promise<Profile | undefined> {
```
Comment (kibibites) [inline src/utils/minecraftProfile.ts]: why does it return a `Proimse`, you already awaited the fetch response

File: src/utils/minecraftProfile.ts [hunk 8]
```diff
@@ -1,30 +1,22 @@
...
+  return userProfile instanceof Error ? undefined : userProfile;
 }
 
-function nameHistory(uuid: string): Name[] | undefined {
-  fetch(`https://api.mojang.com/user/profiles/${uuid}/names`)
-    .then((res) => res.json())
-    .then((data) => (data as Name[]))
-    .catch((_err) => false);
-  return undefined;
+async function nameHistory(uuid: string): Promise<Names | undefined> {
```
Comment (kibibites) [inline src/utils/minecraftProfile.ts]: same as above

File: src/utils/minecraftProfile.ts [hunk 9]
```diff
@@ -1,30 +1,22 @@
...
+  return names instanceof Error ? undefined : names;
 }
 
-function textures(uuid: string): Textures | undefined {
-  fetch(`https://sessionserver.mojang.com/session/minecraft/profile/${uuid}`)
-    .then((res) => res.json())
-    .then((data) => (data as Textures))
-    .catch((_err) => false);
-  return undefined;
+async function textures(uuid: string): Promise<TextureURL | undefined> {
```
Comment (kibibites) [inline src/utils/minecraftProfile.ts]: same as above

File: src/utils/uuid.ts [hunk 10]
```diff
@@ -10,11 +10,9 @@ const format = (uuid: string) => {
   return `${part1}-${part2}-${part3}-${part4}-${part5}`;
 };
 
-function usernameToUUID(username: string): string | undefined {
-  fetch(`https://api.mojang.com/users/profiles/minecraft/${username}`)
-    .then((res) => res.json())
-    .then((data: any) => data.data.id);
-  return undefined;
+async function usernameToUUID(username: string): Promise<string | undefined> {
```
Comment (kibibites) [inline src/utils/uuid.ts:13]: same as above

File: tsconfig.json [hunk 11]
```diff
@@ -9,7 +9,7 @@
     // "disableSourceOfProjectReferenceRedirect": true,  /* Disable preferring source files instead of declaration files when referencing composite projects. */
     // "disableSolutionSearching": true,                 /* Opt a project out of multi-project reference checking when editing. */
     // "disableReferencedProjectLoad": true,             /* Reduce the number of projects loaded automatically by TypeScript. */
-
+
```
Comment (kibibites) [inline tsconfig.json:12]: bruh

**Comment by kibibites** [comment]
atob is deprecated btw

**Comment by kibibites** [comment]
this is finally done

**Pull request merged**